contracts/MechMarketplace.sol

```diff
@@ -489,7 +506,161 @@ contract MechMarketplace is IErrorsMarketplace {
+        emit MarketplaceDeliver(priorityMech, msg.sender, mechDelivery.requester, requestId, deliveryData);
+
+        _locked = 1;
+    }
+
+    /// @dev Verifies provided request hash against its signature.
+    /// @param requester Requester address.
+    /// @param requestHash Request hash.
+    /// @param signature Signature bytes associated with the signed request hash.
+    function _verifySignedHash(address requester, bytes32 requestHash, bytes memory signature) internal view {
+        // Check for zero address
+        if (requester == address(0)) {
+            revert ZeroAddress();
+        }
+
+        // Check for the signature length
+        if (signature.length != 65) {
+            revert IncorrectSignatureLength(signature, signature.length, 65);
+        }
+
+        // Decode the signature
+        uint8 v = uint8(signature[64]);
+        bytes32 r;
+        bytes32 s;
+
+        address recRequester;
+        // Special case of approved hash, where the address of the requester is encoded into r
+        if (v == 4) {
+            recRequester = address(uint160(uint256(r)));
+
+            // Hashes have been pre-approved by the requester via a separate tx, see requesterApproveHash() function
+            if (!mapRequesterApprovedHashes[recRequester][requestHash]) {
+                revert HashNotApproved(recRequester, requestHash, signature);
+            }
+
+            return;
+        }
+
+        // Check EIP-1271 signature validity if requester is a contract
+        if (requester.code.length > 0) {
+            if (ISignatureValidator(requester).isValidSignature(requestHash, signature) == MAGIC_VALUE) {
+                return;
+            } else {
+                revert HashNotValidated(requester, requestHash, signature);
+            }
+        }
+
+        // For the correct ecrecover() function execution, the v value must be set to {0,1} + 27
+        // Although v in a very rare case can be equal to {2,3} (with a probability of 3.73e-37%)
+        // If v is set to just 0 or 1 when signing by the EOA, it is most likely signed by the ledger and must be adjusted
+        if (v < 4) {
+            // In case of a non-contract, adjust v to follow the standard ecrecover case
+            v += 27;
+        }
+        // solhint-disable-next-line no-inline-assembly
+        assembly {
+            r := mload(add(signature, 32))
+            s := mload(add(signature, 64))
+        }
+
+        // Case of ecrecover with the request hash for EOA signatures
+        recRequester = ecrecover(requestHash, v, r, s);
+
+        // Final check is for the requester address itself
+        if (recRequester != requester) {
+            revert WrongRequesterAddress(recRequester, requester);
+        }
+    }
+
+    /// @dev Approves request hash for requester address.
+    /// @param hash Provided request hash to approve.
+    function requesterApproveHash(bytes32 hash) external {
```

Comment:
what is the usecase for this?

Reply:
If called via Safe, for example, such that the path of not going via `isValidSignature()` is executed. We had a long debate with Andrey about it, and ultimately decided to leave this scenario as it's supported by Safe.

Reply:
Had another long discussion about this, let's remove this ambiguous scenario. Correct Safe supports `isValidSignature()` to follow the standard, so we need to consider the general case of a contract following the EIP-1271 workflow for signature verification.